custom-api/store_server.py

```diff
@@ -0,0 +1,296 @@
+    )
+
+    stores = None
+    if statement_response.status.state == StatementState.SUCCEEDED:
+        stores = statement_response.result.data_array
+    else:
+        print(statement_response.status)
+
+    response = {
+        'state': str(statement_response.status.state.name),
+        'stores': stores 
+    }
+    
+    return jsonify(response)
+
+#
+# Endpoint to list sales for a specific store.
+#
+# This invokes the Statement Execution API in asynchronous mode, returning an ID the client can use to poll on status.
+# Once the status is SUCCEEDED, the number of result chunks is returned to the caller. Then, in a parallel, the caller
+# can get external links for each chunk and download the data from cloud storage.
+#
+# Arguments:
+#   store_id - Required ID of the store supplied in the URL.
+#   request_id - Optional ID used to poll for status on a request.
+#   chunk_index - Optional chunk index that can be used to get an external link to data for a successful request.
+#
+# Return Values:
+#   request_id - Returned for all successfully accepted requests. Used this ID to poll for completion status.
+#   state - Returned on the initial request and status polls. Can be PENDING, RUNNING, SUCCEEDED, FAILED, CANCELED,
+#           or CLOSED.
+#   chunk_count - Returned for successful requests, indicating how many chunks of sales data there are for the store.
+#   link - A pre-signed URL returned when a request_id and chunk_index are supplied as arguments.
+#   
+@app.route("/api/1.0/stores/<string:store_id>/sales", methods=["GET"])
+def list_sales_async(store_id):
+    # If a request ID is not supplied, execute a new list request.
+    request_id = request.args.get('request_id')
+    chunk_index = request.args.get('chunk_index')
+    if request_id is None:
+        # Get the format from the request. It can be ‘CSV’, ‘JSON_ARRAY’, or ‘ARROW_STREAM’
+        format = request.args.get('format')
+        row_limit = request.args.get('row_limit')
+        response = execute_list_sales_request(store_id, format, row_limit)
+        
+    # If there is no chunk index, poll for status.
+    elif chunk_index is None:
+        response = get_list_sales_status(request_id)
+
+    # Otherwise return the pre-signed URL for the specified chunk.
+    else:
+        response = get_list_sales_chunk(request_id, chunk_index)
+        
+    return jsonify(response)
+
+# Invoke the Statement Execution API in asynchronous mode, returning an ID so the client can poll on status.
+def execute_list_sales_request(store_id, format, row_limit):
+    # Use parameters to prevent SQL injection via the store ID string.
+    parameters = [
+        StatementParameterListItem(name='store_id', value=store_id, type="INT")
+    ]
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.LIST_SALES],
+        format = Format[format],
+        disposition = Disposition.EXTERNAL_LINKS,
+        wait_timeout = "0s",
```

Comment:
Yeah. It does seem about double the page load time, even for me. ~800ms for 100 stores and ~1200-1600ms for 20 sales. And it is the 2 extra round trips. We could optimize away one of them.

I was trying to not have the disposition as part of the API, but could add that. It would be less code sharing on the FE though.